gravitee-apim-gateway/gravitee-apim-gateway-http/src/main/java/io/gravitee/gateway/jupiter/http/vertx/VertxHttpServerRequest.java

```diff
@@ -237,41 +238,40 @@ public Single<Buffer> bodyOrEmpty() {
 
     @Override
     public void body(Buffer buffer) {
-        this.chunks = chunks.compose(upstream -> Flowable.just(buffer));
+        cacheChunks(this.chunks.compose(upstream -> Flowable.just(buffer)));
     }
 
     @Override
     public Completable onBody(MaybeTransformer<Buffer, Buffer> onBody) {
         // Reduce all the chunks then apply the transformation.
-        final Maybe<Buffer> body = chunks.reduce(Buffer::appendBuffer).compose(onBody);
-        cacheChunks(body.toFlowable(), true);
-
-        return chunks.ignoreElements();
+        cacheChunks(chunks().reduce(Buffer::appendBuffer).compose(onBody).toFlowable());
+        return this.chunks.ignoreElements();
     }
 
     @Override
     public Flowable<Buffer> chunks() {
+        if (this.chunks == null) {
+            this.chunks = Flowable.empty();
+        }
         return chunks;
     }
 
     @Override
     public void chunks(final Flowable<Buffer> chunks) {
-        this.chunks = chunks.compose(upstream -> chunks);
+        if (this.chunks == null) {
+            cacheChunks(chunks);
+        } else {
+            cacheChunks(this.chunks.compose(upstream -> chunks));
+        }
```

Comment:
which case are you talking about ? we will cache only if someone subscribe to it, if we do not cache here if someone then call chunks(), the underlying flowable won't be cached and if someone else resubscribe to it we will reconsume all.

Reply:
Unfortunately as long as we will have v3 policies to support, there will be multiple subscriptions because of policies working at request_content or response_content.

Saying that, we know that we must add cache to avoid re-subscriptions but it is definitely not something we want to keep once v3 aren't supported anymore (we will make sure to have only one subscription).

I'm more about trying to optimize the caching to avoid those that could be useless (and it will become more and more useless as things progress on policy migration). We know that the sub-sequent subscriptions to the request's chunks are made by the `PolicyAdapter` that subscribes to `request.body()` prior call a policy working on the `request_content` scope. This case is already covered by the `body()` method but is no more efficient as it is now caching each time the `body()` method is called. If we have a chain of N policies that just read the body without altering it, we will cache the body N times.

We also know that there is a unique subscription to the `request.chunks()` which is made by the `InvokerAdapter` to send the chunks to the underlying v3 invoker. No trouble if something did something to alter the body before cause we will have only one subscribe to the final `request.chunk()`. No problem also if there are multiple subscription on the `request.body()` before the `InvokerAdapter` is called because `request.body()` takes care of caching also.

As a conclusion, we should pay a special attention to the caching request's body and try to optimize it as much as possible as it could have unexpected performance issues in some complex situations.

Happy to discuss further (and maybe see how we could implement test on this part).

Reply:
I have do some change on both Request and Reponse to share how they work, and also only caching when access is made on the body or chunks and not when setting it. Please have a new look.

Reply:
Nice idea the `AbstractHttpChunks` 👍 
I see that you've completely removed the cache on the body() where I was more thinking about keeping it but optimize it by caching only for the first time (like it was before).